v2/protocol/http/protocol_retry.go

```diff
@@ -0,0 +1,123 @@
+package http
+
+import (
+	"context"
+	"errors"
+	"go.uber.org/zap"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/cloudevents/sdk-go/v2/binding"
+	cecontext "github.com/cloudevents/sdk-go/v2/context"
+	"github.com/cloudevents/sdk-go/v2/protocol"
+)
+
+func (p *Protocol) do(ctx context.Context, req *http.Request) (binding.Message, error) {
+	params := cecontext.RetriesFrom(ctx)
+
+	switch params.Strategy {
+	case cecontext.BackoffStrategyConstant, cecontext.BackoffStrategyLinear, cecontext.BackoffStrategyExponential:
+		return p.doWithRetry(ctx, params)(req)
+	case cecontext.BackoffStrategyNone:
+		fallthrough
+	default:
+		return p.doOnce(req)
+	}
+}
+
+func (p *Protocol) doOnce(req *http.Request) (binding.Message, protocol.Result) {
+	resp, err := p.Client.Do(req)
+	if err != nil {
+		return nil, protocol.NewReceipt(false, "%w", err)
+	}
+
+	var result protocol.Result
+	if resp.StatusCode/100 == 2 {
+		result = protocol.ResultACK
+	} else {
+		result = protocol.ResultNACK
+	}
+
+	return NewMessage(resp.Header, resp.Body), NewResult(resp.StatusCode, "%w", result)
+}
+
+func (p *Protocol) doWithRetry(ctx context.Context, params *cecontext.RetryParams) func(*http.Request) (binding.Message, error) {
+	return func(req *http.Request) (binding.Message, error) {
+		params := params
+		then := time.Now()
+		retry := 0
+		results := make([]protocol.Result, 0)
+
+		for {
+			msg, result := p.doOnce(req)
+
+			// Fast track common case.
+			if protocol.IsACK(result) {
+				return msg, NewRetriesResult(result, retry, then, results)
+			}
+
+			// Try again?
+			//
+			// Make sure the error was something we should retry.
+
+			//{
+			//	var sErr *syscall.Errno
+			//	if errors.As(result, &sErr) {
```

Comment:
I think we should not care of the underlying syscall error, because we already check the `*url.Error`

Reply:
removed.